### ISSUE TITLE
Concurrent Modification Exception

### ISSUE THREAD

**Issue by GoogleCodeExporter** [vadbch/orika#97]
```
I got a ConcurrentModificationExceptio in DefaultMapperFactory:637 (Orika 
1.4.1):

ma.glasnost.orika.MappingException: Error encountered while mapping for the 
following inputs: 
rawSource=kn._int.mdm.cip.ws.maintenance.v6.GetCountryTemplateRequest@103d828
sourceClass=class kn._int.mdm.cip.ws.maintenance.v6.GetCountryTemplateRequest
destinationClass=class com.kn.cip.ws.service.requests.GetCountryTemplateRequest
    at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:687)
    at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:656)
    at com.kn.cip.ws.service.mapper.WebServiceMapper.mapGetCountryTemplateRequest(WebServiceMapper.java:806)
    at com.kn.cip.ws.service.MaintenanceWS.getCountryTemplate(MaintenanceWS.java:560)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:39)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:25)
    at java.lang.reflect.Method.invoke(Method.java:597)
    at com.sun.xml.ws.api.server.InstanceResolver$1.invoke(InstanceResolver.java:246)
    at com.sun.xml.ws.server.InvokerTube$2.invoke(InvokerTube.java:146)
    at com.sun.xml.ws.server.sei.EndpointMethodHandler.invoke(EndpointMethodHandler.java:257)
    at com.sun.xml.ws.server.sei.SEIInvokerTube.processRequest(SEIInvokerTube.java:93)
    at com.sun.xml.ws.api.pipe.Fiber.__doRun(Fiber.java:598)
    at com.sun.xml.ws.api.pipe.Fiber._doRun(Fiber.java:557)
    at com.sun.xml.ws.api.pipe.Fiber.doRun(Fiber.java:542)
    at com.sun.xml.ws.api.pipe.Fiber.runSync(Fiber.java:439)
    at com.sun.xml.ws.server.WSEndpointImpl$2.process(WSEndpointImpl.java:243)
    at com.sun.xml.ws.transport.http.HttpAdapter$HttpToolkit.handle(HttpAdapter.java:471)
    at com.sun.xml.ws.transport.http.HttpAdapter.handle(HttpAdapter.java:244)
    at com.sun.xml.ws.transport.http.servlet.ServletAdapter.handle(ServletAdapter.java:135)
    at com.sun.xml.ws.transport.http.servlet.WSServletDelegate.doGet(WSServletDelegate.java:129)
    at com.sun.xml.ws.transport.http.servlet.WSServletDelegate.doPost(WSServletDelegate.java:160)
    at com.sun.xml.ws.transport.http.servlet.WSServlet.doPost(WSServlet.java:75)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:637)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:290)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at com.kn.cip.app.presentation.PerformanceMeasurementWebServiceFilter.doFilter(PerformanceMeasurementWebServiceFilter.java:127)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at com.kn.cip.app.filter.FlexFaultFilter.doFilter(FlexFaultFilter.java:44)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:185)
    at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:159)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:235)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:233)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:191)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:470)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:127)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:102)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:109)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:293)
    at org.apache.coyote.http11.Http11Processor.process(Http11Processor.java:859)
    at org.apache.coyote.http11.Http11Protocol$Http11ConnectionHandler.process(Http11Protocol.java:602)
    at org.apache.tomcat.util.net.JIoEndpoint$Worker.run(JIoEndpoint.java:489)
    at java.lang.Thread.run(Thread.java:662)
Caused by: java.util.ConcurrentModificationException
    at java.util.LinkedList$ListItr.checkForComodification(LinkedList.java:761)
    at java.util.LinkedList$ListItr.next(LinkedList.java:696)
    at ma.glasnost.orika.impl.DefaultMapperFactory.getRegisteredMapper(DefaultMapperFactory.java:637)
    at ma.glasnost.orika.impl.DefaultMapperFactory.existsRegisteredMapper(DefaultMapperFactory.java:613)
    at ma.glasnost.orika.impl.MapperFacadeImpl.normalizeSourceType(MapperFacadeImpl.java:105)
    at ma.glasnost.orika.impl.MapperFacadeImpl.resolveMappingStrategy(MapperFacadeImpl.java:155)
    at ma.glasnost.orika.impl.MapperFacadeImpl.map(MapperFacadeImpl.java:675)
    ... 47 more

As in the nature of such problems, it just happens rarely. Orika is used there 
to map requests and responses of WebServices within a Tomcat. So it is used in 
different threads at the same time. As a result of 'unlocky' timing, one thread 
might search a mapper while another has just created a new mapper. I guess that 
that happened in my example.

When I have a look at the code, I see a LinkedList that is used unprotected 
from multiple threads.

If my analysis is true, you have several possibilities to solve the problem (I 
dont know, which is the fastes as runtime):
* Use a CopyOnWriteArrayList in SortedCollection 
* Create a copy of the ArrayList in SortedCollection.iterator and return the 
iterator on the copy.

Orika is a great product and we really like to use it. Its running stable on 
most of our servers but we sometimes unreproducable crashes on production 
system where afterwards the whole tomcat is not able to run. The effect above 
is the first problem we found on development system and is eventually a hint to 
the production system.

I hope, I can help you with this issue to improve the quality of this allready 
good tool. Thanks!
```

Original issue reported on code.google.com by `o...@olafpanz.de` on 18 Apr 2013 at 6:12

**Comment by GoogleCodeExporter** [comment]
```
Thanks for your support and issue report :)

Your analysis is relevant, and point where lay the issue. As you mentioned 
there is several possibilities, number of reads are much bigger than 
modification so copying on read may cost a lot. 
I think a volatile reference  + immutable (create new collection on each 
modification) collection may be good tradeoff


```

Original comment by `elaat...@gmail.com` on 19 Apr 2013 at 11:02
- Changed state: **Started**

**Comment by GoogleCodeExporter** [comment]
```
One more question: Do you need the ability of adding an element more then once? 
If not, you may use a ConcurrentSkipListSet instead of the SortedCollection. If 
you must go with sorted collection, you can use Collections.binarySearch () to 
evaulate the insert index. The method needs a sorted list and find the index 
with binary search (log2(n) not n/2, so in a list with 1000 elements, it needs 
10 checks, while the existing impl needs 500 checks on everage)
```

Original comment by `o...@olafpanz.de` on 22 Apr 2013 at 8:41

**Comment by GoogleCodeExporter** [comment]
```
[deleted comment]
```

**Comment by GoogleCodeExporter** [comment]
```
The main purpose of this SortedCollection is to keep order of registered 
mappers, there is no need to have more than one occurence of each element, so 
ConcurrentSkipListSet may be a good solution, I was just wondering if it is 
available since 1.5 ? I need to verify 

```

Original comment by `elaat...@gmail.com` on 22 Apr 2013 at 5:09

**Comment by GoogleCodeExporter** [comment]
```
The CopyOnWriteArrayList exists since JDK 1.5 while ConcurrentSkipListSet came 
with 1.6. So if Orika needs to be compatible to 1.5, I would use the 
CopyOnWriteArrayList in combination with Collections.binarySearch(..) for 
insert (exists since 1.2). 
```

Original comment by `o...@olafpanz.de` on 23 Apr 2013 at 6:04

**Comment by GoogleCodeExporter** [comment]
```
I think maybe we can use the jsr166x dependency to get ConcurrentSkipListSet in 
jdk 1.5...but I don't think we can use this particular class for 
OrderedCollection, as the sorting behavior is somewhat unique -- there are no 
duplicate elements in the collection, but some of the elements may sort as 
equal; most data structures with built-in sorting wouldn't allow this.

Sorting as "equal" in this particular case means the elements are unrelated -- 
which is the case when two mappers have unrelated class hierarchies; in this 
case, the new mapper is inserted after all of the other mappers in the list it 
is "equal" to with regards to sort. This allows us to make sure as we iterate 
through the sorted mappers that the more specific ones (for a given class 
hierarchy) appear before the more generic ones.

But I think we can use ConcurrentLinkedDeque (also available in 166x) to get 
the job done...

```

Original comment by `matt.deb...@gmail.com` on 1 May 2013 at 7:47

**Comment by GoogleCodeExporter** [comment]
```
A fix has been uploaded which uses ConcurrentSkipListMap as the backing and 
should support concurrent access now. Additional test cases to follow, but feel 
free to check out the latest version of the 1.4.2-SNAPSHOT and let us know your 
results.
```

Original comment by `matt.deb...@gmail.com` on 6 May 2013 at 5:36

**Comment by GoogleCodeExporter** [comment]
```
Fixed in 1.4.2 release
```

Original comment by `matt.deb...@gmail.com` on 20 May 2013 at 2:47
- Changed state: **Fixed**
